agent/agent.go

```diff
@@ -1159,9 +1150,32 @@ func (a *Agent) Shutdown() error {
 		a.logger.Println("[WARN] agent: could not delete pid file ", pidErr)
 	}
 
-	a.logger.Println("[INFO] agent: shutdown complete")
+	// stage 2: shutdown HTTP and DNS endpoints and trigger final shutdown notification
+
+	go func() {
+		// Stop all API endpoints
+		for _, srv := range a.dnsServers {
+			a.logger.Printf("[INFO] agent: Stopping DNS server %s (%s)", srv.Server.Addr, srv.Server.Net)
+			srv.Shutdown()
+		}
+		for _, srv := range a.httpServers {
+			a.logger.Printf("[INFO] agent: Stopping %s server %s", strings.ToUpper(srv.proto), srv.Addr)
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+			defer cancel()
+			srv.Shutdown(ctx)
+			if ctx.Err() == context.DeadlineExceeded {
+				a.logger.Printf("[WARN] agent: Timeout stopping %s server %s", strings.ToUpper(srv.proto), srv.Addr)
+			}
+		}
+		a.logger.Println("[INFO] agent: Waiting for endpoints to shut down")
+		a.wgServers.Wait()
+		a.logger.Print("[INFO] agent: Endpoints down")
+
+		a.logger.Println("[INFO] agent: shutdown complete")
+		close(a.shutdownCh)
+	}()
+
```

Comment:
It looks like it has to be this way (not blocking on the http server shutdowns) so that in the Leave endpoint this function can return and finish serving the connection to the client, then shut down the http servers afterward (because they're doing a graceful shutdown and waiting for connections to finish being served). If we shut them down before Leave returns, we get the EOF.